### PR TITLE
Classifier based on Siamese Network to Classify Alzheimer's Disease on ADNI Dataset - 47571840

### PR DESCRIPTION
**Pull request by afrahhhardian**
**Name**: Adzkia Aisyah Afrah Hardian
**Student Number**: s4757184

**Task Chosen**: 7. Create a classifier based on a Siamese network [10] to classify either Alzheimer’s disease (normal and AD) of the ADNI brain data set having an accuracy of around 0.8 on the test set [Hard Difficulty].

**Objective**: This pull request implements a Siamese Network that will be trained to generate embeddings of images of the ADNI dataset. A classifier (fully connected layers) is added at the end of the trained Siamese Network to classify Alzheimer's Disease on the dataset.

**Model**: The backbone architecture of the Siamese network is based on PyTorch's ResNet-18 with modifications. The classifier's architecture consists of fully connected layers to do classification tasks. A trained Siamese model sits at the end of the classifier with its weights frozen. 

**Usage**: Dependencies and reproducibility of code are covered in the README.md. Most importantly, the ADNI dataset should already exist in your local repository and modify the data paths in the `train.py` and `predict.py` if needed.

**Performance**: The best classifier reached 0.721 accuracy on the test set.

**Files added**: 

- `README.md`  report of the project consisting of an overview of the Siamese network, description of the ADNI dataset, pre-processing steps, training results, testing results, dependencies, and references. 

- `dataset.py` provides functions and classes to load the ADNI dataset for training and testing the Siamese and classifier, including the functions to pre-process data. 

- `modules.py` consists of the Siamese and classifier module classes, including the ContrastiveLoss class. 
- `train.py` trains the Siamese and classifier models and evaluates the classifier on the test set. 
- `predict.py` loads existing trained Siamese and trained classifier and do inference using both models of one input image.
-  `/images_for_readme` folder contains images displayed in the README.md

### CODE REVIEW

**Comment by nathasha-naranpanawa** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Hard

Readme: Very good
- Clear flow of information
- Algortihm and training details explained well
- Relevant plots for loss, accuracy provided
- Presents results for several different model combinations and hyperparameter settings

Commit messages: OK

Code: 
- design is good
- sufficiently commented
- algorithm seems to be implemented correctly

Functionality/Performance: 
- siamese seem to perform well
- classifier is very unstable and strange on the validation set, but it seems no actions have been taken to resolve it
- classifier achieves only 72% on the test set

General comments:
- algorithm solves the problem apporpriately although performance is somewhat poor

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable/good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
